miner/currency_blocklist.go

```diff
@@ -0,0 +1,136 @@
+package miner
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const hours uint64 = 60 * 60
+
+var evictionTimeoutSeconds uint64 = 2 * hours
+
+type AddressBlocklist struct {
+	mux        *sync.RWMutex
+	currencies map[common.Address]*types.Header
+	// fee-currencies blocked at headers with an older timestamp
+	// will get evicted when evict() is called
+	headerEvictionTimeoutSeconds uint64
+	oldestHeader                 *types.Header
+}
+
+func NewAddressBlocklist() *AddressBlocklist {
+	return &AddressBlocklist{
+		mux:                          &sync.RWMutex{},
+		currencies:                   map[common.Address]*types.Header{},
+		headerEvictionTimeoutSeconds: evictionTimeoutSeconds,
+		oldestHeader:                 nil,
+	}
+}
+
+func (b *AddressBlocklist) FilterAllowlist(allowlist common.AddressSet, latest *types.Header) common.AddressSet {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+
+	filtered := common.AddressSet{}
+	for a := range allowlist {
+		if !b.isBlocked(a, latest) {
+			filtered[a] = struct{}{}
+		}
+	}
+	return filtered
+}
+
+func (b *AddressBlocklist) IsBlocked(currency common.Address, latest *types.Header) bool {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+
+	return b.isBlocked(currency, latest)
+}
+
+func (b *AddressBlocklist) Remove(currency common.Address) bool {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	h, ok := b.currencies[currency]
+	if !ok {
+		return false
+	}
+	delete(b.currencies, currency)
```

Comment:
If we deleted the last currency from `b.currencies` here, does it mean that the call on line 62 wouldn't actually do anything and exit early (lines 84-86), meaning that the header would stay there indefinitely? Adding new currencies (with more recent header times) won't overwrite this value and on the next call would be considered evicted.

Reply:
Yes, you are right. Setting the header to `nil` in this case will fix this.
 I will also add some tests for the blocklist.